Add version_family flag to lock to a given stemcell "line"

ljfranklin opened the pull request:
- Useful for locking to a given hotfix line of stemcells

Signed-off-by: Brian Cunnie bcunnie@pivotal.io

[#133033179](https://www.pivotaltracker.com/story/show/133033179)

ljfranklin committed:
Add version_family flag to lock to a given stemcell "line"

- Useful for locking to a given hotfix line of stemcells

Signed-off-by: Brian Cunnie <bcunnie@pivotal.io>

[#133033179](https://www.pivotaltracker.com/story/show/133033179)

zachgersh commented:
@ljfranklin haven't been ignoring this I promise. Took a quick glance at the code. Was wondering if there was a way to leverage the semver package more? Can we pick a different format for our filter that allows us to not have the case statement / more filtering code?

ljfranklin commented:
@zachgersh Are you suggesting swapping out `version_family: 3263` for something like `version_range: >=3263 <3264`? If so, that feels more error prone for the user and locks the resource pretty tightly to that specific semver library.

zachgersh commented:
@ljfranklin I figured out a way to do this that is simplified. I am planning on implementing it but I haven't had the time :/

I have a new structure that I'd like to propose:

``` yaml
version_family:
  major: 3264
  tracking: (( minor || patch ))
```

I don't really ever see a need for someone to track only major releases so I have made a format that leads you to tracking minor / patch releases for the major you've specified. Does this work for you use cases?

ljfranklin commented:
@zachgersh I'm not sure I understand what `tracking: patch` would do. If I want all stemcells in the "3264.X" line, I would say `tracking: minor`. If I wanted only patch updates (e.g. "3264.4.X"), how would I specify which minor version to pin to? Would I say `major: 3264, minor: 4, tracking: patch`?

ljfranklin commented:
@zachgersh apparently BOSH has recently added a feature that lets you do the following:

``` yaml
stemcells:
- os: ubuntu-trusty
  version: 3147.latest
```

Should we change this PR to the following for consistency:

``` yaml
resources:
- name: aws-stemcell
  type: bosh-io-stemcell
  source:
    name: bosh-aws-xen-ubuntu-trusty-go_agent
    stemcell_version: (( latest || 3147 || 3147.latest ))
```

ljfranklin commented:
@zachgersh bump, what do you think about the new syntax show above ^^^^^?

zachgersh commented:
@ljfranklin sorry for the delay on this - I still haven't thought of a good syntax or way to actually implement this in the resource. Really, the cleanest solution would be to try to push this down into the semver library we are using so we can isolate what "family" you want to use.

I would be open to pairing on that and then trying to redo this PR. Would you want to do that?

ljfranklin commented:
@zachgersh absolutely, just ping me when you have time and we'll try to make it happen. For the syntax, if BOSH now supports the "3147.latest" format it seems like this resource should use the same syntax for consistency. That said, I couldn't find any docs on that version syntax. Pinging @cppforlife and @dpb587-pivotal to see if that syntax is supported / documented somewhere.

pivotal-issuemaster commented:
@ljfranklin Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/bosh-io-stemcell-resource&pullRequestId=15) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

pivotal-issuemaster commented:
@ljfranklin Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/bosh-io-stemcell-resource&pullRequestId=15)!

ljfranklin commented:
Hey @zachgersh, we're back from the break and ready to finish this out. We can put a pair on updating the PR to the using the syntax mentioned [here](https://github.com/concourse/bosh-io-stemcell-resource/pull/15#issuecomment-259495587). Sounded like you were a bit hesitant to use that syntax, but it seems the most consistent with BOSH. We'll wait for you to give the go ahead before picking up the story.

zachgersh commented:
@ljfranklin I'll merge it if you finish it out - I don't want to hold you all up any longer :D

dpb587-pivotal commented:
Oops; `.latest` was implemented in [#119505701](https://www.pivotaltracker.com/story/show/119505701) with logic in [73378f7](https://github.com/cloudfoundry/bosh/commit/73378f73d4d8bddcb0f64de80e929b885692c24d?diff=split#diff-d440e9d4fe159277666e0ea8436eb2c7R18).

Commit:
Use version_family as defined in:

```
(( latest || 3147 || 3147.latest ))
```

[#133222913](https://www.pivotaltracker.com/story/show/133222913)

Signed-off-by: Lyle Franklin <lfranklin@pivotal.io>
Signed-off-by: Chris Dutra <cdutra@pivotal.io>
Signed-off-by: Brian Cunnie <bcunnie@pivotal.io>

cunnie commented:
@zachgersh Hi Zach, could you please look over our PR when you have time and see if it meets your expectations? If so, merge, if not, let us know & we'll be happy to fix.

Short version:
```yaml
resources:
- name: aws-stemcell
  type: bosh-io-stemcell
  source:
    name: bosh-aws-xen-ubuntu-trusty-go_agent
    version_family: (( latest || 3147 || 3147.latest ))
```

zachgersh commented:
@cunnie / @ljfranklin this is in now - I am going to do some slight refactoring but you should be able to use it.
